playwright-test/harness.ts

```diff
@@ -0,0 +1,125 @@
+import {
+    test as base,
+    chromium,
+    type BrowserContext,
+    type Worker,
+    type Route,
+    type Page,
+    type Request,
+} from "@playwright/test";
+import path from "path";
+import fs from "fs/promises";
+
+function getManifestVersion() {
+    return process.env.npm_lifecycle_event === "playwright-mv3" ? 3 : 2;
+}
+
+async function routeLocalResources(route: Route) {
+    const url = new URL(route.request().url());
+    const localPath = path.join(__dirname, "data", url.pathname);
+    try {
+        const body = await fs.readFile(localPath);
+        console.log("request served from disk", route.request().url());
+        route.fulfill({ status: 200, body });
+    } catch (e) {
+        console.log("request served from network", route.request().url());
+        route.continue();
+    }
+}
+
+export const mockAtb = {
+    majorVersion: 364,
+    minorVersion: 2,
+    version: 'v364-2'
+};
+
+// based off example at https://playwright.dev/docs/chrome-extensions#testing
+export const test = base.extend<{
+    manifestVersion: 2 | 3;
+    context: BrowserContext;
+    backgroundPage: Page | Worker;
+    routeExtensionRequests: (url: string | RegExp, handler: (route: Route, request: Request) => any) => Promise<void>;
+    backgroundNetworkContext: Page | BrowserContext;
+}>({
+    manifestVersion: getManifestVersion(),
+    context: async ({ manifestVersion }, use) => {
+        const extensionPath =
+            manifestVersion === 3 ? "build/chrome-mv3/dev" : "build/chrome/dev";
+        const pathToExtension = path.join(__dirname, "..", extensionPath);
+        const context = await chromium.launchPersistentContext("", {
+            headless: false,
+            args: [
+                `--disable-extensions-except=${pathToExtension}`,
+                `--load-extension=${pathToExtension}`,
+            ],
+        });
+        // intercept extension install page and use HAR
+        context.on("page", (page) => {
+            console.log("page", page.url());
+            if (page.url().includes("duckduckgo.com/extension-success")) {
+                page.routeFromHAR(path.join(__dirname, "data", "postinstall.har"), {
```

Comment:
I'm curious, how are these HAR files generated? (Sorry, I've not looked through the docs yet.)

Also, perhaps the HAR files should live in a separate directory (e.g. `./data/har-responses`) if we're going to keep adding more? Also, perhaps we should follow a convention with their naming, like start with the URL and then have a description? I'm thinking something like `duckduckgo.com/extension-success - postinstall.har` might be good. WDYT?

Reply:
Playwright has a built in generator tool. I added a comment for how to create this one: https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/0296948f8e0c83b575b2a20f79e965e0877a0f12/playwright-test/harness.js#L49

Good suggestion on a naming convention.

Reply:
Ace thanks :+1: Would you also mind adding a README.md to the har directory? Just to quickly explain what the files are for and to include that example command.